### PR TITLE
[5.6] Update queue worker memory usage to use the "real" amount of memory used

### PR DESCRIPTION
**Pull request by matthew-muscat**
This update makes a small change to the queue worker memory calculation logic to use the "real usage" calculation, which will include memory that may be allocated outside of the application scripts being processed by the queue, but still being allocated to the PHP thread overall.

The reason for this change is to allow the worker daemon to more accurately calculate the amount of memory allocated to it, improving the "memory" worker's logic to restart the queue worker when it reaches a memory limit defined by the user.

We encountered an issue whereby queue workers were exceeding 800mb of allocated RSS memory to a worker thread, despite having a memory limit in place of 32mb. This was due to a large portion of the memory not being allocated within the bounds of the script, but more so in the system memory (our queues workers process a large number of external HTTP calls, which appear to have a significant amount of memory allocated outside of the value calculated by `get_memory_usage(false)`

Changing the calculation to use `get_memory_usage(true)` is now resulting in the queues restarting around the 120mb allocated memory mark (with a worker limit of 32mb).

Note: Due to the nature of PHP and memory allocation, the number would still be lower than the overall memory that is actually consumed by the worker thread, but this change should result in the calculation being much more closer to the real value, especially when the memory is allocated outside of the script, but still within the php worker thread — all while still maintaining using the internal PHP method call for `get_memory_usage()`.

Example references...
https://github.com/laravel/framework/issues/16783#issuecomment-269476161

### CODE REVIEW

**Comment by matthew-muscat** [comment]
Note — it would be great if we could get this also merged into the 5.5 and 5.7 releases — let me know if we need separate pull requests for this.

**Comment by matthew-muscat** [comment]
I've also updated the Queue Listener class, which has similar logic in place

**Comment by mfn** [comment]
> which will include memory that may be allocated outside of the application scripts being processed by the queue, but still being allocated to the PHP thread overall.

This statement, to my knowledge, is not correct.

But before trying to answer it with my own words, see https://stackoverflow.com/a/2290965/47573
> So, there are two kinds of memory usage:
> - How much memory the engine took from the OS ("real usage")
> - How much of this memory was actually used by the application ("internal usage")

Additionally see also https://stackoverflow.com/a/38707938/47573 .

So in any event, this still _is_ memory allocated solely by PHP and nothing else.

The difference:
- with `false` (the default), it reports only the "actually used parts" of the memory pool
- with `true` it reports the "actually allocated" memory from the OS _but_ it may not yet have used this

ELI5: PHP allocates memory in bigger, always increasing chunks from the OS and then gives them out in the small chunks as used by the script internally.

Question for you: when you're seeing the 800MB RSS, what does `get_memory_usage(true)` actually report in your case?

My slight issue with the PR is that it's possible PHP pre-allocated memory which isn't yet used by the application but already fires the trigger.

**Comment by matthew-muscat** [comment]
> My slight issue with the PR is that it's possible PHP pre-allocated memory which isn't yet used by the application but already fires the trigger.

Understood — although, this would still be memory wired and used by the OS — my understanding is that the purpose of the `memory` flag on the queue worker is to avoid the worker from exceeding desired memory limits.

> Question for you: when you're seeing the 800MB RSS, what does get_memory_usage(true) actually report in your case?

That's an excellent question — unfortunately, as this issue occurred in production, i was unable to debug this at the time
